server/src/test/java/org/elasticsearch/monitor/fs/FsHealthServiceTests.java

```diff
@@ -0,0 +1,185 @@
+            fsHealthService = new FsHealthService(settings, clusterSettings, testThreadPool, env);
+            fsHealthService.new FsHealthMonitor().run();
+            assertEquals(fsHealthService.getHealth(), NodeHealthService.Status.UNHEALTHY);
+        } finally {
+            ThreadPool.terminate(testThreadPool, 500, TimeUnit.MILLISECONDS);
+            disruptFileSystemProvider.injectIOException.set(false);
+        }
+    }
+
+    public void testFailsHealthOnSinglePathIOException() throws IOException {
+        FileSystem current = PathUtils.getDefaultFileSystem();
+        FileSystemIOExceptionProvider disruptFileSystemProvider = new FileSystemIOExceptionProvider(current);
+        PathUtilsForTesting.installMock(disruptFileSystemProvider.getFileSystem(null));
+        final Settings settings = Settings.EMPTY;
+        final ClusterSettings clusterSettings = new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+        TestThreadPool testThreadPool = new TestThreadPool(getClass().getName(), settings);
+        try (NodeEnvironment env = newNodeEnvironment()) {
+            Path[] paths = env.nodeDataPaths();
+            FsHealthService fsHealthService = new FsHealthService(settings, clusterSettings, testThreadPool, env);
+            fsHealthService.new FsHealthMonitor().run();
+            assertEquals(fsHealthService.getHealth(), NodeHealthService.Status.HEALTHY);
+
+            //disrupt file system on single path
+            disruptFileSystemProvider.injectIOException.set(true);
+            disruptFileSystemProvider.restrictPathPrefix(randomFrom(paths).toString());
+            fsHealthService = new FsHealthService(settings, clusterSettings, testThreadPool, env);
+            fsHealthService.new FsHealthMonitor().run();
+            assertEquals(fsHealthService.getHealth(), NodeHealthService.Status.UNHEALTHY);
+            assertEquals(disruptFileSystemProvider.getInjectedPathCount(), 1);
+
+        } finally {
+            ThreadPool.terminate(testThreadPool, 500, TimeUnit.MILLISECONDS);
+            disruptFileSystemProvider.injectIOException.set(false);
+        }
+    }
+
+    private static class FileSystemIOExceptionProvider extends FilterFileSystemProvider {
+
+        AtomicBoolean injectIOException = new AtomicBoolean();
+        AtomicInteger injectedPaths = new AtomicInteger();
+
+        private String pathPrefix = "/";
+
+        FileSystemIOExceptionProvider(FileSystem inner) {
+            super("disrupt_fs_health://", inner);
+        }
+
+        public void restrictPathPrefix(String pathPrefix){
+            this.pathPrefix = pathPrefix;
+        }
+
+        public int getInjectedPathCount(){
+            return injectedPaths.get();
+        }
+
+        @Override
+        public FileChannel newFileChannel(Path path, Set<? extends OpenOption> options, FileAttribute<?>... attrs) throws IOException {
+            if (injectIOException.get()){
```

Comment:
I'd prefer that we also check the behaviour on writes and syncs, effectively to assert that we do detect failures for both of those activities too.